grudge/op.py

```diff
@@ -92,11 +92,68 @@
 )
 
 
+# {{{ common derivative "kernels"
+
+def _single_axis_derivative_kernel(
+        actx, out_discr, in_discr, get_diff_mat, inv_jac_mat, xyz_axis, vec,
+        *, metric_in_matvec):
+    return DOFArray(
+        actx,
+        data=tuple(
+            # r for rst axis
+            actx.einsum("rej,rij,ej->ei" if metric_in_matvec else "rei,rij,ej->ei",
+                        ijm_i[xyz_axis],
+                        get_diff_mat(
+                            actx,
+                            out_element_group=out_grp,
+                            in_element_group=in_grp
+                        ),
+                        vec_i,
+                        arg_names=("inv_jac_t", "ref_stiffT_mat", "vec", ),
+                        tagged=(FirstAxisIsElementsTag(),))
+
+            for out_grp, in_grp, vec_i, ijm_i in zip(
+                out_discr.groups, in_discr.groups, vec,
+                inv_jac_mat)))
+
+
+def _gradient_kernel(actx, out_discr, in_discr, get_diff_mat, inv_jac_mat, vec,
+        *, metric_in_matvec):
+    per_group_grads = [
+        # r for rst axis
+        # x for xyz axis
+        actx.einsum("xrej,rij,ej->xei" if metric_in_matvec else "xrei,rij,ej->xei",
+                    ijm_i,
+                    get_diff_mat(
+                        actx,
+                        out_element_group=out_grp,
+                        in_element_group=in_grp
+                    ),
+                    vec_i,
+                    arg_names=("inv_jac_t", "ref_stiffT_mat", "vec"),
+                    tagged=(FirstAxisIsElementsTag(),))
+        for out_grp, in_grp, vec_i, ijm_i in zip(
+            out_discr.groups, in_discr.groups, vec,
+            inv_jac_mat)]
+
+    return make_obj_array([
+            DOFArray(
+                actx, data=tuple([pgg_i[xyz_axis] for pgg_i in per_group_grads]))
+            for xyz_axis in range(out_discr.ambient_dim)])
```

Comment:
Maybe emphasize here in this function (as a comment or dosctring, nothing too complicated) that this routine is more or less supporting two types of differentiation (weak and strong --- depending on what gets passed in as `get_diff_mat` and `metric_in_matvec`.

Reply:
Agreed, see 3987a0a.